Add tutorial on events and triggers (#3505)
sxs-collaboration/spectre: docs/Tutorials/EventsAndTriggers.md

\cond NEVER
Distributed under the MIT License.
See LICENSE.txt for details.
\endcond

# Events and triggers {#tutorial_events_and_triggers}

The "events and triggers" infrastructure (and the related "events and
dense triggers") provide some control of SpECTRE executable execution
from the input file contents.  It does not generally have major

> effects on the flow of the algorithm, but is primarily used for

Reviewer: \ref dev_guide_parallelization_core_algorithm algorithm

> observations and minor adjustments.

Reviewer 2: I think it'd be good to clarify the meaning of the term 'observation' which we use here and after since this is one of the tutorial docs. Some phrases like 'writing/recording simulation data' would be good.

Reviewer: I think a \ref would be good here to the base for an observer (if there is one).

Author: I don't think we have anything appropriate to point to here.  There's a tutorial on visualization, but that's mostly focused on things like ParaView, and we have a dev guide on the really low-level observation code, but you probably don't even need to understand that if you are writing an observer.  (I have added some words to the text here.)

> ### What is an event?

Reviewer: I'm personally not a fan of the "question and answer" type of documentation. Also I don't think it's used anywhere else in spectre (if it is used, it's sparse). For consistency I think we should keep it as
- Event
- Trigger
- Dense Trigger
- Input File Syntax

An \ref Event "event" is an input-file-creatable object that performs

> some task, frequently an observation.  With the exception of the \ref

Reviewer 2: Can you provide some concrete examples of event in casual words? Like 'write data to a file',  'record error norms', or 'exit program'. (I cannot think of other good examples for now)

Author: This is sort of what I was going for with the rest of the paragraph, but it is not ordered in the way that makes sense to think about the concepts, so I'll redo that.

Events::Completion "Completion" event, they cannot have direct effects
on the algorithm, although they can have indirect effects through
sending messages.  (An example of such an event is \ref

> Events::ChangeSlabSize "ChangeSlabSize".)  In most cases, however, their

Reviewer: This was confusing upon a first read. If I understood this correctly, clarify that by "cannot have direct effects on the algorithm" you mean that essentially events can't add actions to the pdal which is specified in the metavariables. But you can (potentially) change things in the databox or cache (like the slab size).

Author: Kind of the opposite, actually.  They can't change the DataBox.  I guess they can technically do cache updates, but doing those from an array element seems generally unlikely to work.  They cannot do anything except send messages.  Hopefully the reword here helps.

effects on program execution are ephemeral, causing some data to be
written to a file or standard output but not changing anything about
the simulation state.

### What is a trigger?

A \ref Trigger "trigger" is an input-file-creatable object that
controls when events are run.  They are checked periodically, once per

> slab in an evolution executable, and once per iteration in an elliptic

Reviewer 2: \ref Slab may help a bit

solve.  At each check, a trigger may fire, causing its associated
events to run, or not.  Triggers must always be written to be
independent of element-specific state such as the local values of the
system variables, so that a trigger will give a consistent result over

> the entire domain at each check.

Reviewer: This sentence was a bit wordy and confusing, specifically the phrasing of "independent of element-specific state".

Author: Reversed the clauses.  Any better?

### What is a dense trigger?

A \ref DenseTrigger "dense trigger" is similar to a trigger, but uses

> time-stepper dense output to provide more precise control of the time

Reviewer 2: Q. What is the time-stepper dense output?

Reviewer: A \ref would be helpful here as well.

Author: Added "(i.e., interpolation or extrapolation)", but I don't think we have any documentation on this topic.

that it fires.  As a time-related feature, dense triggers are only
applicable to evolution executables.  Dense triggers fire at a precise
set of times, independent of the time-step or slab state, by
interpolating the evolved variables to the requested time before
running the associated events.  As with normal triggers, dense
triggers are required to fire consistently over the entire domain.

### What's with the weird input file syntax?

The `%EventsAndTriggers` (and `%EventsAndDenseTriggers`) sections of
the input file indicate structure using a collection of `?`, `:`, and
`-` characters.  These are standard, if perhaps somewhat obscure, YAML
syntax.  They are defining a map from triggers to lists of events.
The `?:` syntax defines an expanded key-value pair, which allows the
key to be a complicated type, unlike the common `Key: Value` syntax.
The `-` is the expanded form of a list, which allows entries to have
complicated types.  Frequently only one event is associated with a
trigger, so there will only be a single `-` after the `:`, but
additional events introduced by more `-` characters can appear
afterwards:

\snippet PlaneWave1DEventsAndTriggersExample.yaml multiple_events

> 

Reviewer 2: It would help a lot if there is a description on this snippet based on the contents previously introduced -- which one stands for an event, which one stands for a trigger, and which lines are options for an event or a trigger, etc.

Author: Done, but unfortunately I can't link one of the events because of doxygen limitations.

> If an event and trigger both took no options, the syntax
> 
> \snippet PlaneWave1DEventsAndTriggersExample.yaml no_arguments
> 
> could be written in the more familiar form
> 
> \snippet PlaneWave1DEventsAndTriggersExample.yaml no_arguments_single_line
> 
> but we recommend always using the expanded form for consistency.

Reviewer: If you don't want users to use this format for triggers/events with no options, then don't put it in. It'll only confuse them more. What about this for this part:
```suggestion
If an event and trigger both took no options, the recommended syntax is still

\snippet PlaneWave1DEventsAndTriggersExample.yaml no_arguments
```

Author: I was trying to teach something beyond how to cut and paste.  If you think it is really a bad idea I can leave the whole thing out, but I don't see any reason to mention this case if we're not using to to explain something.

Reviewer: This was more of a nitpick so I think it's ok as is.

\cond NEVER
Distributed under the MIT License.
See LICENSE.txt for details.
\endcond

# Events and triggers {#tutorial_events_and_triggers}

The "events and triggers" infrastructure (and the related "events and
dense triggers") provide some control of SpECTRE executable execution
from the input file contents by running user-specified code on the
elements.  It does not generally have major effects on the flow of the
\ref dev_guide_parallelization_core_algorithm "algorithm", but is
primarily used for observations, such as writing volume data to H5
files, and minor simulation adjustments.

### Events

An \ref Event "event" is an input-file-creatable object that performs
some task that (with the exception of the \ref Events::Completion

> "Completion" event), does not directly affect the execution of the

Reviewer: [nitpick] No comma after the parentheses.

algorithm on the element.  The effects of events are limited to
sending messages.  The most commonly used events (such as the \ref
dg::Events::ObserveErrorNorms "ObserveErrorNorms" event) send data to
be written to disk, but have no long-term effects on the simulation

> state.  Others, (such as \ref Events::ChangeSlabSize "ChangeSlabSize")

Reviewer: [nitpck] No comma

can have indirect effects when \ref dev_guide_parallelization_actions
"actions" react to the messages they send.

### Triggers

A \ref Trigger "trigger" is an input-file-creatable object that
controls when events are run.  They are checked periodically, once per
Slab in an evolution executable, and once per iteration in an elliptic
solve.  At each check, a trigger may fire, causing its associated

> events to run, or not.  Triggers must give a consistent result over

Reviewer: [nitpick] "At each check a trigger may fire, causing its associated events to run or not."
